File: src/main/java/org/antop/java/chapter9/Math.java
package org.antop.java.chapter9;

import java.io.IOException;

public class Math {

    public static <T extends Number> double oneHalf(T value) {
        return value.doubleValue() / 2;
    }

    public static <T extends Comparable<T>> T max(T first, T second) {
        return (first.compareTo(second) > 0) ? first : second;
    }

    public static <T extends CharSequence & Appendable> void ensureTrailingPeriod(T seq) {
        if (seq == null) return;
        int length = seq.length();
        if (length > 0 && seq.charAt(length - 1) != '.') {
            try {
                seq.append('.');
            } catch (IOException e) {
                // do nothing
            }
        }
    }

    public static void main(String[] args) {
        double half = oneHalf(4L);
        System.out.println("half = " + half);

        System.out.println(max("kotlin", "java"));

        StringBuilder sb = new StringBuilder("Hello World");
        ensureTrailingPeriod(sb);
    }

}

File: src/main/kotlin/org/antop/kotlin/chapter9/Math.kt
package org.antop.kotlin.chapter9

fun <T : Number> oneHalf(value: T): Double {
    return value.toDouble() / 2.0
}

fun <T : Comparable<T>> max(first: T, second: T): T {
    return if (first > second) first else second
}

fun <T> ensureTrailingPeriod(seq: T)
        where T : CharSequence, T : Appendable {
    if (!seq.endsWith('.')) {
        seq.append('.')
    }
}

fun main() {
    val half = oneHalf(4)
    println("half = $half")

    // 다른 타입이기 때문에 컴파일 에러
    // println(max("kotlin", 42))

    println(max("kotlin", "java"))

    val sb = StringBuilder("hello")
    ensureTrailingPeriod(sb)
    println("sb = $sb")
}
